test/test_common/simulated_time_system.cc

```diff
@@ -308,34 +279,40 @@ Thread::CondVar::WaitStatus SimulatedTimeSystemHelper::waitFor(Thread::MutexBasi
   return Thread::CondVar::WaitStatus::Timeout;
 }
 
-MonotonicTime
-SimulatedTimeSystemHelper::alarmTimeLockHeld(Alarm* alarm) ABSL_NO_THREAD_SAFETY_ANALYSIS {
-  // We disable thread-safety analysis as the compiler can't detect that
-  // alarm_->timeSystem() == this, so we must be holding the right mutex.
-  ASSERT(&(alarm->timeSystem()) == this);
-  return alarm->time();
-}
-
 void SimulatedTimeSystemHelper::alarmActivateLockHeld(Alarm* alarm) ABSL_NO_THREAD_SAFETY_ANALYSIS {
   // We disable thread-safety analysis as the compiler can't detect that
   // alarm_->timeSystem() == this, so we must be holding the right mutex.
   ASSERT(&(alarm->timeSystem()) == this);
   alarm->activateLockHeld();
 }
 
-int64_t SimulatedTimeSystemHelper::nextIndex() {
-  absl::MutexLock lock(&mutex_);
-  return index_++;
-}
-
 void SimulatedTimeSystemHelper::addAlarmLockHeld(
     Alarm* alarm, const std::chrono::microseconds& duration) ABSL_NO_THREAD_SAFETY_ANALYSIS {
   ASSERT(&(alarm->timeSystem()) == this);
-  alarm->setTimeLockHeld(monotonic_time_ + duration);
-  alarms_.insert(alarm);
+  ASSERT(alarms_.size() == alarm_registrations_map_.size());
+  ASSERT(alarm_registrations_map_.find(alarm) == alarm_registrations_map_.end());
+
+  auto insert_result = alarms_.insert({monotonic_time_ + duration, random_source_.random(), alarm});
+  ASSERT(insert_result.second);
+  alarm_registrations_map_.emplace(alarm, insert_result.first);
+
+  // Sanity check that the parallel data structures used for alarm registration have the same number
+  // of entries.
+  ASSERT(alarms_.size() == alarm_registrations_map_.size());
```

Comment:
I'm a little unclear why the structure of this had to change so much, as opposed to simply replacing `index_` with `randomness_` in the structure.

Does that provide some flexibility previously lacking?

Reply:
I wanted to recompute randomness on every insert, not just pick a random value at creation time.  Also, having the timeout outside the timer is closer to what I'm used to.

Reply:
That doesn't seem at all hard to do in the current structure; we need to know if an alarm is in the map, and can't mutate a map-key while in the map. I think this is just a question of mutating the randomness_ field immediately prior to insertion, which happens only in `addAlarmLockHeld`.

I'm not that wedded to the structure that is there now, but it has these 2 benefits:
 * easier to review due to less changes in code
 * not having 2 parallel structures to keep in sync.

Reply:
I'm really not a fan of the current structure.  Mutating the alarm feels too subtle, brittle and error prone.  I can offer to do a no-op refactor to an alternate structure first.

Reply:
Actually if you are signing up to own this I am totally down with the rewrite :)

No need for an intermediate refactor.